airflow/cli/commands/connection.py

```diff
@@ -0,0 +1,463 @@
+        login=conn.login,
+        password=conn.password,
+        host=conn.host,
+        port=conn.port,
+        schema=conn.schema,
+        extra=conn.extra,
+    )
+
+
+def _format_connections(conns: List[Connection], file_format: str, serialization_format: str) -> str:
+    if serialization_format == 'json':
+        serializer_func = lambda x: json.dumps(_connection_to_dict(x))
+    elif serialization_format == 'uri':
+        serializer_func = Connection.get_uri
+    else:
+        raise SystemExit(f"Received unexpected value for `--serialization-format`: {serialization_format!r}")
+    if file_format == '.env':
+        connections_env = ""
+        for conn in conns:
+            connections_env += f"{conn.conn_id}={serializer_func(conn)}\n"
+        return connections_env
+
+    connections_dict = {}
+    for conn in conns:
+        connections_dict[conn.conn_id] = _connection_to_dict(conn)
+
+    if file_format == '.yaml':
+        return yaml.dump(connections_dict)
+
+    if file_format == '.json':
+        return json.dumps(connections_dict, indent=2)
+
+    return json.dumps(connections_dict)
+
+
+def _is_stdout(fileio: io.TextIOWrapper) -> bool:
+    return fileio.name == '<stdout>'
+
+
+def _valid_uri(uri: str) -> bool:
+    """Check if a URI is valid, by checking if both scheme and netloc are available"""
+    uri_parts = urlparse(uri)
+    return uri_parts.scheme != '' and uri_parts.netloc != ''
+
+
+@connections.command('export')
+@click.pass_context
+@click_file_output
+@click_verbose
+@click_file_format
+@click_format
+@click_serialization_format
+def connections_export(ctx, output, verbose, file_format, format, serialization_format):
+    """Exports all connections to a file"""
+    file_is_stdout = False
+    file_output = False
+    file_formats = ['.yaml', '.json', '.env']
+    if format:
+        warnings.warn("Option `--format` is deprecated.  Use `--file-format` instead.", DeprecationWarning)
+    if format and file_format:
+        raise SystemExit('Option `--format` is deprecated.  Use `--file-format` instead.')
+    default_format = '.json'
+    provided_file_format = None
+    if format or file_format:
+        provided_file_format = f".{(format or file_format).lower()}"
+    
+    if output.strip() == '-':
+        file_is_stdout = True
+    else:
+        try:
+            open(output, 'w')
+            file_output = True
+        except OSError:
+            raise SystemExit('Option `--output` is invalid.  Provide correct file path')
+        
+
+    if file_is_stdout:
+        filetype = provided_file_format or default_format
+    elif provided_file_format:
+        filetype = provided_file_format
+    else:
+        filetype = Path(output).suffix
+        filetype = filetype.lower()
+        if filetype not in file_formats:
+            raise SystemExit(
+                f"Unsupported file format. The file must have the extension {', '.join(file_formats)}."
+            )
```

Comment:
This:

```python
        try:
            open(output, 'w')
            file_output = True
        except OSError:
            raise SystemExit('Option `--output` is invalid.  Provide correct file path')
```

will _always_ leak a file descriptor and never close the file while the command runs. That's probably not a huge issue for a command that will likely only have a lifetime of a few seconds, but it's not a clean solution and I would recommend against this using pattern at all in any project. Python has a [better way to handle this](https://docs.python.org/3/tutorial/inputoutput.html#reading-and-writing-files) using [context managers](https://book.pythontips.com/en/latest/context_managers.html#context-managers):

> It is good practice to use the [with](https://docs.python.org/3/reference/compound_stmts.html#with) keyword when dealing with file objects. The advantage is that the file is properly closed after its suite finishes, even if an exception is raised at some point. Using with is also much shorter than writing equivalent [try](https://docs.python.org/3/reference/compound_stmts.html#try)-[finally](https://docs.python.org/3/reference/compound_stmts.html#finally) blocks

The good news is, if we use Click's `click.File` type specifier, I think we can get rid of pretty much all of this code:

```suggestion
```

(There will be some additional cleanups necessary to get this to work and pass the lint check.)

Reply:
@blag i used `open` without any context-manager to check if the file exists. That way I can check both if the path is valid and file_exists, I am not sure if we could do that with context-manager, to open file and do nothing. Please point to me if i am missing to understand any point.

I will check if adding `click.File` will get rid of this code part. I am not sure if it will ensure passed file path is valid and exists/creatable. I will confirm it and make the change

Reply:
It's fine to try to open a file to validate that it exists. The problem with the original code is that it opens a file but never closes it. This keeps an unreachable file descriptor resource around, which is wasteful and is generally bad form. We would have to do something like this to properly handle it:

```python
try:
    file = open(...)
except OSError:
    ...
else:
    file.close()
```

but that's exactly what the `open()` context manager already does. We can use the context manager like so:

```python
try:
    with open(..., 'w'):
        pass  # do nothing with it
except OSError:
    raise SystemExit('Option `--output` is invalid.  Provide correct file path')
```

But we should be able to use [`click.File(mode='w', encoding='UTF-8')`](https://click.palletsprojects.com/en/8.1.x/parameters/) here, which will open the file for us in writing mode with the UTF-8 encoding will allow us to use the same API for writing to stdout/stderr and an actual file, and will also temporarily validate the file for us _and_ close it if necessary:

> The lazy flag controls if the file should be opened immediately or upon first IO. The default is to be non-lazy for standard input and output streams as well as files opened for reading, lazy otherwise. When opening a file lazily for reading, **it is still opened temporarily for validation**, but will not be held open until first IO. lazy is mainly useful when opening for writing to avoid creating the file until it is needed.

So if we properly leverage Click here then we can get rid of lots of code. 🤘

Reply:
Cool i will try this @blag Thanks

Reply:
> It's fine to try to open a file to validate that it exists. The problem with the original code is that it opens a file but never closes it. This keeps an unreachable file descriptor resource around, which is wasteful and is generally bad form. We would have to do something like this to properly handle it:
> 
> ```python
> try:
>     file = open(...)
> except OSError:
>     ...
> else:
>     file.close()
> ```
> 
> but that's exactly what the `open()` context manager already does. We can use the context manager like so:
> 
> ```python
> try:
>     with open(..., 'w'):
>         pass  # do nothing with it
> except OSError:
>     raise SystemExit('Option `--output` is invalid.  Provide correct file path')
> ```
> 
> But we should be able to use [`click.File(mode='w', encoding='UTF-8')`](https://click.palletsprojects.com/en/8.1.x/parameters/) here, which will open the file for us in writing mode with the UTF-8 encoding will allow us to use the same API for writing to stdout/stderr and an actual file, and will also temporarily validate the file for us _and_ close it if necessary:
> 
> > The lazy flag controls if the file should be opened immediately or upon first IO. The default is to be non-lazy for standard input and output streams as well as files opened for reading, lazy otherwise. When opening a file lazily for reading, **it is still opened temporarily for validation**, but will not be held open until first IO. lazy is mainly useful when opening for writing to avoid creating the file until it is needed.
> 
> So if we properly leverage Click here then we can get rid of lots of code. 🤘

@blag Yes this works :) Click is really powerful.

Reply:
@blag I have removed most parts of this code. Can you check and tell if I can still improve?